src/nodes.coffee

```diff
@@ -2247,13 +2252,19 @@ exports.Assign = class Assign extends Base
 
       restElements
 
-    # Cache the value for reuse with rest elements
-    [@value, valueRef] = @value.cache o
+    # Cache the value for reuse with rest elements.
+    if @value.shouldCache()
```

Comment:
I don't think this is needed any more? I was having no problems without any of the `valueRefTemp` stuff.

Reply:
Also - in the PR description you mention that object literals were not being cached, however they do appear to be on `2` currently:

http://coffeescript.org/v2/#try:%7Ba%2C%20r...%7D%20%3D%20%7Ba%3A1%2C%20b%3A2%7D

Reply:
@connec how did you get rid of the `valueRefTemp` references? How did you revise the
```coffee
restElements = traverseRest @variable.base.properties, valueRefTemp
```
line?

@zdenko please ignore the polyfill discussion, that’s been taken care of in a separate PR. Do you have time to address these last few notes so that we can merge this?

Reply:
@GeoffreyBooth yes, I'll take a look a.s.a.p.

Reply:
> in the PR description you mention that object literals were not being cached

@connec I was wrong regarding the caching. It was probably my mistake in the code.
`::shouldCache()` and `valueRefTemp` are used to assign object literal `value` to `ref` variable, e.g. `{p: {m, q..., t = {obj...}}, r...} = c: 2` => `({p: {m, t = Object.assign({}, obj)}} = ref = { c: 2})`.
The reason for use `valueRefTemp` here, are cases  where  there are no rest elements in the destructuring, e.g. `{a = {b...}} = c:1` => `({a = Object.assign({}, b)} = {c: 1})`.

So, tests will pass if you remove `valueRefTemp` and use `@value` as a parameter for the `traverseRest()`, only the compiled code will be different.

With `traverseRestTemp` and `::shouldCache()`
```coffeescript
{p: {m, q..., t = {obj...}}, r...} = c: 2
###
({
  p: {m, t = Object.assign({}, obj)}
} = ref = {
  c: 2
});
q = objectWithoutKeys(ref.p, ['m', 't']);
r = objectWithoutKeys(ref, ['p']);
###
```

and without
```coffeescript
{p: {m, q..., t = {obj...}}, r...} = c: 2
###
({
  p: {m, t = Object.assign({}, obj)}
} = {
  c: 2
});
q = objectWithoutKeys({
  c: 2
}.p, ['m', 't']);
r = objectWithoutKeys({
  c: 2
}, ['p']);
###
```

Reply:
Sorry @zdenko, I think I must be misunderstanding something. If I remove the `valueTempRef` stuff:

```diff
diff --git a/src/nodes.coffee b/src/nodes.coffee
index 2e30bfe4..5e331f34 100644
--- a/src/nodes.coffee
+++ b/src/nodes.coffee
@@ -2253,16 +2253,12 @@ exports.Assign = class Assign extends Base
       restElements
 
     # Cache the value for reuse with rest elements.
-    if @value.shouldCache()
-      valueRefTemp = new IdentifierLiteral o.scope.freeVariable 'ref', reserve: false
-    else
-      valueRefTemp = @value.base
+    [@value, valueRef] = @value.cache o
 
     # Find all rest elements.
-    restElements = traverseRest @variable.base.properties, valueRefTemp
+    restElements = traverseRest @variable.base.properties, valueRef
     return no unless restElements and restElements.length > 0
 
-    [@value, valueRef] = @value.cache o
     result = new Block [@]
 
     for restElement in restElements

```

And run the example you gave:

```sh
echo '{p: {m, q..., t = {obj...}}, r...} = c: 2' | bin/coffee -bcs
```

I get the desired output:

```js
// Generated by CoffeeScript 2.0.0-beta4
var m, q, r, ref, ref1, t,
  objectWithoutKeys = function(o, ks) { var res = {}; for (var k in o) ([].indexOf.call(ks, k) < 0 && {}.hasOwnProperty.call(o, k)) && (res[k] = o[k]); return res; };

({
  p: {m, t = Object.assign({}, obj)}
} = ref1 = ref = {
  c: 2
});
q = objectWithoutKeys(ref.p, ['m', 't']);
r = objectWithoutKeys(ref, ['p']);
```

Reply:
*Edit:* in fact, I see that that output 'double-cached' `{c:2}` (as `ref` and `ref1`). I guess this is because `@value` is assigned even if the function returns. The following patch avoids the double-cache:

```diff
diff --git a/src/nodes.coffee b/src/nodes.coffee
index 2e30bfe4..081c2a71 100644
--- a/src/nodes.coffee
+++ b/src/nodes.coffee
@@ -2253,16 +2253,13 @@ exports.Assign = class Assign extends Base
       restElements
 
     # Cache the value for reuse with rest elements.
-    if @value.shouldCache()
-      valueRefTemp = new IdentifierLiteral o.scope.freeVariable 'ref', reserve: false
-    else
-      valueRefTemp = @value.base
+    [cachedValue, valueRef] = @value.cache o
 
     # Find all rest elements.
-    restElements = traverseRest @variable.base.properties, valueRefTemp
+    restElements = traverseRest @variable.base.properties, valueRef
     return no unless restElements and restElements.length > 0
 
-    [@value, valueRef] = @value.cache o
+    @value = cachedValue
     result = new Block [@]
 
     for restElement in restElements

```

Sorry to be fussy about this, but I'd rather not proliferate manual caching code (e.g. `new IndentifierLiteral o.scope.freeVariable`) when `Node::cache` could be used instead.

Reply:
I've tried that before. With this approach `ref` variable is still declared in the scope by call from  `::cache` method: 
```coffeescript
if complex
  ref = new IdentifierLiteral o.scope.freeVariable 'ref'
  sub = new Assign ref, this
```

Check the full output:
```coffeescript
{a = {b...}} = c:1
###
var a, ref;
       ^^^
({a = Object.assign({}, b)} = {
  c: 1
});
###

{p: {m, q..., t = {obj...}}, r...} = c: 2
###
var m, q, r, ref, ref1, t,
                  ^^^^
  objectWithoutKeys = function(o, ks) { var res = {}; for (var k in o) ([].indexOf.call(ks, k) < 0 && {}.hasOwnProperty.call(o, k)) && (res[k] = o[k]); return res; };

({
  p: {m, t = Object.assign({}, obj)}
} = ref = {
  c: 2
});
q = objectWithoutKeys(ref.p, ['m', 't']);
r = objectWithoutKeys(ref, ['p']);
###

```

Reply:
Aha, good catch 😢 

In that case, this LGTM!